File: src/main/java/com/milaboratory/mixcr/cli/CommandSlice.java
/*
 * Copyright (c) 2014-2022, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli;

import cc.redberry.pipe.CUtils;
import cc.redberry.pipe.OutputPort;
import cc.redberry.pipe.util.FlatteningOutputPort;
import com.milaboratory.mixcr.basictypes.*;
import com.milaboratory.mixcr.util.Concurrency;
import gnu.trove.map.hash.TIntIntHashMap;
import gnu.trove.set.hash.TLongHashSet;
import io.repseq.core.VDJCLibraryRegistry;
import picocli.CommandLine.Command;
import picocli.CommandLine.Option;
import picocli.CommandLine.Parameters;

import java.nio.file.Paths;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;
import java.util.concurrent.atomic.AtomicLong;

import static com.milaboratory.mixcr.cli.CommandSlice.SLICE_COMMAND_NAME;
import static com.milaboratory.util.TempFileManager.smartTempDestination;

@Command(name = SLICE_COMMAND_NAME,
        sortOptions = true,
        separator = " ",
        description = "Slice ClnA file.")
public class CommandSlice extends MiXCRCommand {
    static final String SLICE_COMMAND_NAME = "slice";

    @Parameters(description = "data.[vdjca|clns|clna]", index = "0")
    public String in;

    @Parameters(description = "data_sliced", index = "1")
    public String out;

    @Option(description = "List of read (for .vdjca) / clone (for .clns/.clna) ids to export.",
            names = {"-i", "--id"})
    public List<Long> ids = new ArrayList<>();

    @Override
    protected List<String> getInputFiles() {
        return Collections.singletonList(in);
    }

    @Override
    protected List<String> getOutputFiles() {
        return Collections.singletonList(out);
    }

    @Override
    public void run0() throws Exception {
        Collections.sort(ids);

        switch (IOUtil.extractFileType(Paths.get(in))) {
            case VDJCA:
                sliceVDJCA();
                break;
            case CLNS:
                throwValidationException("Operation is not yet supported for Clns files.");
                break;
            case CLNA:
                sliceClnA();
                break;
            default:
                throwValidationException("Not supported file type.");
        }
    }

    void sliceVDJCA() throws Exception {
        TLongHashSet set = new TLongHashSet(ids);

        try (VDJCAlignmentsReader reader = new VDJCAlignmentsReader(in);
             VDJCAlignmentsWriter writer = new VDJCAlignmentsWriter(out)) {
            writer.header(reader);
            for (VDJCAlignments alignments : CUtils.it(reader)) {
                if (set.removeAll(alignments.getReadIds()))
                    writer.write(alignments);
                if (set.isEmpty())
                    break;
            }
            writer.writeFooter(reader.reports(), null);
        }
    }

    void sliceClnA() throws Exception {
        try (ClnAReader reader = new ClnAReader(in, VDJCLibraryRegistry.getDefault(), Concurrency.noMoreThan(4));
             ClnAWriter writer = new ClnAWriter(out,
                     smartTempDestination(out, "", false))) {

            // Getting full clone set
            CloneSet cloneSet = reader.readCloneSet();

            // old clone id -> new clone id
            TIntIntHashMap idMapping = new TIntIntHashMap();

            long newNumberOfAlignments = 0;

            // Creating new cloneset
            List<Clone> clones = new ArrayList<>();
            int i = 0;
            List<OutputPort<VDJCAlignments>> allAlignmentsList = new ArrayList<>();
            for (Long cloneId_ : ids) {
                int cloneId = (int) ((long) cloneId_);
                newNumberOfAlignments += reader.numberOfAlignmentsInClone(cloneId);
                Clone clone = cloneSet.get(cloneId);
                idMapping.put(clone.getId(), i);
                clones.add(clone.setId(i).resetParentCloneSet());
                OutputPort<VDJCAlignments> als = reader.readAlignmentsOfClone(cloneId);
                final int ii = i;
                allAlignmentsList.add(() -> {
                    VDJCAlignments al = als.take();
                    if (al == null)
                        return null;
                    return al.withCloneIndex(ii);
                });
                i++;
            }

            CloneSet newCloneSet = new CloneSet(clones, cloneSet.getUsedGenes(), cloneSet.getInfo(), cloneSet.getOrdering());

            OutputPort<VDJCAlignments> allAlignmentsPortRaw = new FlatteningOutputPort<>(CUtils.asOutputPort(allAlignmentsList));
            AtomicLong idGen = new AtomicLong();
            OutputPort<VDJCAlignments> allAlignmentsPort = () -> {
                VDJCAlignments al = allAlignmentsPortRaw.take();
                if (al == null)
                    return null;
                return al.setAlignmentsIndex(idGen.getAndIncrement());
            };

            writer.writeClones(newCloneSet);

            writer.collateAlignments(allAlignmentsPort, newNumberOfAlignments);

            writer.writeFooter(reader.reports(), null);
            writer.writeAlignmentsAndIndex();
        }
    }
}

File: src/main/kotlin/com/milaboratory/mixcr/cli/CommandSlice.kt
/*
 * Copyright (c) 2014-2023, MiLaboratories Inc. All Rights Reserved
 *
 * Before downloading or accessing the software, please read carefully the
 * License Agreement available at:
 * https://github.com/milaboratory/mixcr/blob/develop/LICENSE
 *
 * By downloading or accessing the software, you accept and agree to be bound
 * by the terms of the License Agreement. If you do not want to agree to the terms
 * of the Licensing Agreement, you must not download or access the software.
 */
package com.milaboratory.mixcr.cli

import cc.redberry.pipe.OutputPort
import cc.redberry.pipe.util.filter
import cc.redberry.pipe.util.flatten
import cc.redberry.pipe.util.forEach
import cc.redberry.pipe.util.it
import cc.redberry.pipe.util.map
import cc.redberry.pipe.util.use
import com.milaboratory.app.InputFileType
import com.milaboratory.app.ValidationException
import com.milaboratory.mixcr.basictypes.ClnAReader
import com.milaboratory.mixcr.basictypes.ClnAWriter
import com.milaboratory.mixcr.basictypes.ClnsReader
import com.milaboratory.mixcr.basictypes.ClnsWriter
import com.milaboratory.mixcr.basictypes.Clone
import com.milaboratory.mixcr.basictypes.CloneSet
import com.milaboratory.mixcr.basictypes.IOUtil
import com.milaboratory.mixcr.basictypes.IOUtil.MiXCRFileType.CLNA
import com.milaboratory.mixcr.basictypes.IOUtil.MiXCRFileType.CLNS
import com.milaboratory.mixcr.basictypes.IOUtil.MiXCRFileType.SHMT
import com.milaboratory.mixcr.basictypes.IOUtil.MiXCRFileType.VDJCA
import com.milaboratory.mixcr.basictypes.VDJCAlignments
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsReader
import com.milaboratory.mixcr.basictypes.VDJCAlignmentsWriter
import com.milaboratory.mixcr.trees.SHMTreesReader
import com.milaboratory.mixcr.trees.SHMTreesWriter
import com.milaboratory.mixcr.util.Concurrency
import com.milaboratory.util.TempFileManager
import com.milaboratory.util.exhaustive
import gnu.trove.map.hash.TIntIntHashMap
import gnu.trove.set.hash.TLongHashSet
import io.repseq.core.VDJCLibraryRegistry
import picocli.CommandLine.ArgGroup
import picocli.CommandLine.Command
import picocli.CommandLine.Mixin
import picocli.CommandLine.Option
import picocli.CommandLine.Parameters
import java.nio.file.Path
import java.util.concurrent.atomic.AtomicLong
import kotlin.io.path.readLines

@Command(
    name = CommandSlice.SLICE_COMMAND_NAME,
    description = ["Slice vdjca|clns|clna|shmt file."]
)
class CommandSlice : MiXCRCommandWithOutputs() {
    @Parameters(
        description = ["Input data to filter by ids"],
        paramLabel = "data.(vdjca|clns|clna|shmt)",
        index = "0"
    )
    lateinit var input: Path

    @Parameters(
        description = ["Output file with filtered data"],
        paramLabel = "data_sliced.(vdjca|clns|clna|shmt)",
        index = "1"
    )
    lateinit var out: Path

    @ArgGroup(
        exclusive = true,
        multiplicity = "1",
        order = OptionsOrder.main + 10_100
    )
    lateinit var idsOptions: IdsFilterOptions

    @Mixin
    lateinit var useLocalTemp: UseLocalTempOption

    class IdsFilterOptions {
        @Option(
            description = ["List of read (for .vdjca) / clone (for .clns/.clna) / tree (for .shmt) ids to export."],
            names = ["-i", "--id"],
            required = true,
            paramLabel = "<id>",
            order = 1
        )
        var ids: List<Long>? = null

        @Option(
            description = [
                "File with list of read (for .vdjca) / clone (for .clns/.clna) / tree (for .shmt) ids to export.",
                "Every id on separate line"
            ],
            names = ["--ids-file"],
            required = true,
            paramLabel = "<path>",
            order = 2
        )
        var fileWithIds: Path? = null
    }

    private val ids: List<Long> by lazy {
        val result = idsOptions.ids ?: idsOptions.fileWithIds!!.readLines().map { it.toLong() }
        result.sorted()
    }

    override val inputFiles
        get() = listOf(input) + listOfNotNull(idsOptions.fileWithIds)

    override val outputFiles
        get() = listOf(out)

    override fun validate() {
        ValidationException.requireTheSameFileType(
            input,
            out,
            InputFileType.VDJCA,
            InputFileType.CLNS,
            InputFileType.CLNA,
            InputFileType.SHMT
        )
    }

    override fun run1() {
        when (IOUtil.extractFileType(input)) {
            VDJCA -> sliceVDJCA()
            CLNS -> sliceClns()
            CLNA -> sliceClnA()
            SHMT -> sliceShmt()
        }.exhaustive
    }

    private fun sliceVDJCA() {
        val set = TLongHashSet(ids)
        VDJCAlignmentsReader(input).use { reader ->
            VDJCAlignmentsWriter(out).use { writer ->
                writer.inheritHeaderAndFooterFrom(reader)
                for (alignments in reader.it) {
                    if (set.removeAll(alignments.readIds)) writer.write(alignments)
                    if (set.isEmpty) break
                }
                writer.setFooter(reader.footer)
            }
        }
    }

    private fun sliceClnA() {
        ClnAReader(input, VDJCLibraryRegistry.getDefault(), Concurrency.noMoreThan(4)).use { reader ->
            ClnAWriter(out.toFile(), TempFileManager.smartTempDestination(out, "", !useLocalTemp.value)).use { writer ->
                // Getting full clone set
                val cloneSet = reader.readCloneSet()

                // old clone id -> new clone id
                val idMapping = TIntIntHashMap()
                var newNumberOfAlignments: Long = 0

                // Creating new cloneset
                val clones = mutableListOf<Clone>()
                val allAlignmentsList = mutableListOf<OutputPort<VDJCAlignments>>()
                for ((i, cloneId_) in ids.withIndex()) {
                    val cloneId = cloneId_.toInt()
                    newNumberOfAlignments += reader.numberOfAlignmentsInClone(cloneId)
                    val clone = cloneSet[cloneId]
                    idMapping.put(clone.id, i)
                    clones.add(clone.withId(i).withTotalCountsReset())
                    allAlignmentsList += reader
                        .readAlignmentsOfClone(cloneId)
                        .map { it.withCloneIndex(i.toLong()) }
                }
                val newCloneSet = CloneSet.Builder(clones, cloneSet.usedGenes, cloneSet.header)
                    .sort(cloneSet.ordering)
                    .recalculateRanks()
                    .calculateTotalCounts()
                    .build()
                val idGen = AtomicLong()
                val allAlignmentsPort = allAlignmentsList
                    .flatten()
                    .map { it.setAlignmentsIndex(idGen.getAndIncrement()) }
                writer.writeClones(newCloneSet)
                writer.collateAlignments(allAlignmentsPort, newNumberOfAlignments)
                writer.setFooter(reader.footer)
                writer.writeAlignmentsAndIndex()
            }
        }
    }

    private fun sliceClns() {
        ClnsReader(input, VDJCLibraryRegistry.getDefault()).use { reader ->
            ClnsWriter(out.toFile()).use { writer ->
                // Getting full clone set
                val cloneSet = reader.readCloneSet()

                // Creating new cloneset
                val clones = mutableListOf<Clone>()
                for ((i, cloneId_) in ids.withIndex()) {
                    val cloneId = cloneId_.toInt()
                    val clone = cloneSet[cloneId]
                    clones.add(clone.withId(i).withTotalCountsReset())
                }
                val newCloneSet = CloneSet.Builder(clones, cloneSet.usedGenes, cloneSet.header)
                    .sort(cloneSet.ordering)
                    .recalculateRanks()
                    .calculateTotalCounts()
                    .build()
                writer.writeCloneSet(newCloneSet)
                writer.setFooter(reader.footer)
            }
        }
    }

    private fun sliceShmt() {
        SHMTreesReader(input, VDJCLibraryRegistry.getDefault()).use { reader ->
            SHMTreesWriter(out).use { writer ->
                writer.copyHeaderFrom(reader)

                writer.treesWriter().use { treesWriter ->
                    reader.readTrees()
                        .filter { it.treeId.toLong() in ids }
                        .forEach { treesWriter.put(it) }
                }

                writer.setFooter(reader.footer)
            }
        }
    }

    companion object {
        const val SLICE_COMMAND_NAME = "slice"
    }
}
